contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -0,0 +1,117 @@
+
+    HotVars internal s_hotVars; // Mixture of config and state, commonly accessed
+
+    mapping(bytes32 => address) internal s_reports;
+
+    constructor(AccessControllerInterface accessController) ConfirmedOwner(msg.sender) {
+        s_accessController = accessController;
+    }
+
+    function splitSignature(
+        bytes memory sig
+    ) public pure returns (bytes32 r, bytes32 s, uint8 v) {
+        require(sig.length == 65, "invalid signature length");
+
+        assembly {
+            /*
+            First 32 bytes stores the length of the signature
+
+            add(sig, 32) = pointer of sig + 32
+            effectively, skips first 32 bytes of signature
+
+            mload(p) loads next 32 bytes starting at the memory address p into memory
+            */
+
+            // first 32 bytes, after the length prefix
+            r := mload(add(sig, 32))
+            // second 32 bytes
+            s := mload(add(sig, 64))
+            // final byte (first byte of the next 32 bytes)
+            v := byte(0, mload(add(sig, 96)))
+        }
+
+        // implicitly return (r, s, v)
+    }
+
+    function splitReport(
+        bytes memory rawReport
+    ) public pure returns (bytes32 workflowId, bytes32 workflowExecutionId) {
+        require(rawReport.length > 64, "invalid report length");
+        assembly {
+            workflowId := mload(add(rawReport, 4))
+            workflowExecutionId := mload(add(rawReport, 36)) // 4 + 32
+        }
+    }
+
+    // send a report to targetAddress
+    // report(workflowRunID, targetAddress, report, signatures)
+    function report(address targetAddress, bytes calldata data, bytes[] calldata signatures) external nonReentrant returns (bool) {
+    	// TODO: maybe forward as foo(report, calldata...) so we can validate sigs separately
+
+        // data is an encoded call with the selector prefixed: (bytes4 selector, bytes report, ...)
+        // we are able to partially decode just the first param, since we don't know the rest
+        bytes memory rawReport = abi.decode(data[4:], (bytes));
+
+        // TODO: we probably need some type of f value config?
+
+        bytes32 hash = keccak256(rawReport);
+
+        // validate signatures
+        for (uint i = 0; i < signatures.length; i++) {
+            // TODO: is libocr-style multiple bytes32 arrays more optimal?
+            (bytes32 r, bytes32 s, uint8 v) = splitSignature(signatures[i]);
+            address signer = ecrecover(hash, v, r, s);
+            // TODO: we need to store oracle cluster similar to aggregator then, to validate valid signer list
+        }
+
+        (bytes32 workflowId, bytes32 workflowExecutionId) = splitReport(rawReport);
+
+        // report was already processed
+        if (s_reports[workflowExecutionId] != address(0)) {
+           return false; 
+        }
+
+        // abi.encodeWithSelector(bytes4 selector, data...)
+        (bool success, bytes memory result) = targetAddress.call(data);
```

Comment:
This can revert, should we call it in try-catch so that s_reports gets updated regardless?

Reply:
`address.call` actually doesn't revert, that's why there's the `bool success` parameter! Had to double check, compiler won't actually accept ` try targetAddress.call(data) returns (bool success, bytes memory result) `